lib/SetAPI/sampleset/SampleSearchUtils.py

```diff
@@ -0,0 +1,124 @@
+        track_total_hits - True/False - Whether to track the total number of hits past the elasticsearch limit
+        """
+        sample_set_ref = params['ref']
+        sort_by, extra_must, start, limit = self._parse_inputs(params)
+
+        (workspace_id, object_id, version) = sample_set_ref.split('/')
+        # we use namespace 'WSVER' for versioned elasticsearch index.
+        ss_id = f'WSVER::{workspace_id}:{object_id}:{version}'
+
+        headers = {"Authorization": self.token}
+        query_data = {
+            "method": "search_objects",
+            "params": {
+                "query": {
+                    "bool": {
+                        "must": [{"term": {"sample_set_ids": ss_id}}] + extra_must
+                    }
+                },
+                "indexes": ["sample"],
+                "from": start,
+                "size": limit,
+                "sort": [{s[0]: {"order": "asc" if s[1] else "desc"}} for s in sort_by]
+            },
+            "jsonrpc": "2.0",
+            "id": str(uuid.uuid4())
+        }
+        if aggs:
+            query_data['params']['aggs'] = aggs
+        if track_total_hits:
+            query_data['params']['track_total_hits'] = True
+        if self.debug:
+            print(f"querying {self.search_url}, with params: {json.dumps(query_data)}")
+        resp = requests.post(self.search_url, headers=headers, data=json.dumps(query_data))
+        if self.debug:
+            print(f"queried {self.search_url} with data {json.dumps(query_data)}")
+        if not resp.ok:
+            raise Exception(f"Not able to complete search request against {self.search_url} "
+                            f"with parameters: \n{json.dumps(query_data)} \nResponse body:\n{resp.text}")
+        respj = resp.json()
+        if self.debug:
+            print(f'respone from query: {respj}')
+        if respj.get('error'):
+            raise Exception(f"Error from Search API with parameters\n{json.dumps(query_data)}\n Response body:\n{respj}")
+        return self._process_sample_set_resp(respj['result'], start, query_data)
+
+    def _process_sample_set_resp(self, resp, start, query):
+        if resp.get('hits'):
+            hits = resp['hits']
+            return {
+                "num_found": int(resp['count']),
+                "start": start,
+                "query": query,
+                # this should handle empty results list of hits, also sort by feature_id
+                "samples": [self._process_sample(h['doc'], h['id']) for h in hits]
+            }
+        # empty list in reponse for hits
+        elif 'hits' in resp:
+            return {
+                "num_found": int(resp['count']),
+                "start": start,
+                "query": query,
+                "samples": []
+            }
+        else:
+            # only raise if no hits found at highest level.
+            raise RuntimeError(f"no 'hits' with params {json.dumps(params)}\n in http response: {resp}")
```

Comment:
Is this a possible error from the service? If it's possible for zero hits, would it return something like `{"hits": {}}` or an error?

Reply:
depends on the type of error. I've had a consistent issue in the past where bad queries don't always result in an error status and instead fail quietly. This condition is to catch those situations